AutoCollection/HeartBeat.ts

```diff
@@ -61,16 +73,88 @@ class HeartBeat {
             properties["appSrv_wsHost"] = process.env.WEBSITE_HOSTNAME || "";
         } else if (process.env.FUNCTIONS_WORKER_RUNTIME) { // Function apps
             properties["azfunction_appId"] = process.env.WEBSITE_HOSTNAME;
-        } 
-        this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
-        callback();
+        } else if (config) {
+            if (this._isVM === undefined) {
+                waiting = true;
+                this._getAzureComputeMetadata(config, () => {
+                    if (this._isVM && Object.keys(this._vmData).length > 0) { // VM
+                        properties["azInst_vmId"] = this._vmData["vmId"] || "";
+                        properties["azInst_subscriptionId"] = this._vmData["subscriptionId"] || "";
+                        properties["azInst_osType"] = this._vmData["osType"] || "";
+                        this._azInst_vmId = this._vmData["vmId"] || "";
+                        this._azInst_subscriptionId = this._vmData["subscriptionId"] || "";
+                        this._azInst_osType = this._vmData["osType"] || "";
+                    }
+                    this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+                    callback();
+                });
+            } else if (this._isVM) {
+                properties["azInst_vmId"] = this._azInst_vmId;
+                properties["azInst_subscriptionId"] = this._azInst_subscriptionId;
+                properties["azInst_osType"] = this._azInst_osType;
+            }
+        }
+        if (!waiting) {
+            this._client.trackMetric({name: Constants.HeartBeatMetricName, value: 0, properties: properties});
+            callback();
+        }
     }
 
     public dispose() {
         HeartBeat.INSTANCE = null;
         this.enable(false);
         this._isInitialized = false;
     }
+
+    private _getAzureComputeMetadata(config: Config, callback: () => void) {
+        const metadataRequestUrl = `${AIMS_URI}?${AIMS_API_VERSION}&${AIMS_FORMAT}`;
+        const requestOptions = {
+            method: 'GET',
+            [AutoCollectHttpDependencies.disableCollectionRequestOption]: true,
+            headers: {
+                "Metadata": "True",
+            }
+        };
+
+        const req = Util.makeRequest(config, metadataRequestUrl, requestOptions, (res) => {
```

Comment:
I think I made this comment on the other PR. There was a bug in OpenCensus Python implementation of heartbeat where an application hangs because the request to azuremetadataservice did not specify a timeout. Might want to do that here too to be safe?

Reply:
I believe this is an async request and therefore the code will not block -- additionally generally XHR requests will fail with a timeout @markwolff Is this your understanding of the Util.makeRequest() as well?

Reply:
Right in Python the request is made via a separate thread. I'm not sure if this is even a problem in NodeJS (this only came up when a user tried to initialize a Django app). So the real question is does `Util.makeRequest()` EVER hang if a timeout is not set? The `requests` library MUST set a timeout or the request itself will wait forever, regardless of whether it is done asynchronously or not.